src/main/java/org/apache/ibatis/builder/annotation/ProviderSqlSource.java

```diff
@@ -54,11 +56,24 @@ public ProviderSqlSource(Configuration configuration, Object provider) {
    * @since 3.4.5
    */
   public ProviderSqlSource(Configuration configuration, Object provider, Class<?> mapperType, Method mapperMethod) {
+    this(configuration, provider, mapperType, mapperMethod, configuration.getDefaultScriptingLanguageInstance());
+  }
+
+  /**
+   * @since 3.4.6
+   */
+  public ProviderSqlSource(Configuration configuration, Object provider, Class<?> mapperType, Method mapperMethod, LanguageDriver languageDriver) {
     String providerMethodName;
     try {
       this.configuration = configuration;
-      this.sqlSourceParser = new SqlSourceBuilder(configuration);
+      //Compatible with velocity and freemarker LanguageDriver
+      if(languageDriver instanceof XMLLanguageDriver){
+        this.languageDriver = languageDriver;
+      } else {
+        this.languageDriver = new XMLLanguageDriver();
```

Comment:
Is there a reason to force languageDriver to XMLLanguageDriver?

Reply:
@h3adache Velocity and Freemarker's languageDriver do not support StaticSqlSource.

When call the `configuration.getdefaultscriptinglanguageinstance( )` method, it is possible to return velocity and freemarker's the languageDriver.

The main purpose is to be compatible with the original method of `ProviderSqlSource`.

Reply:
I would prefer that you leave those to the implementations and not take the current implementations. This is a hidden behavior that might surprise people who try to use this.

Reply:
If I make the other two implementations also support the default behavior, is it ok?

Reply:
Is it possible to change this to the following way?
```java
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface SelectProvider {
  Class<?> type();

  String method();

  Class<? extends SqlSource> providerSqlSource() default ProviderSqlSource.class;
}
```
and
```java
  @SelectProvider(
      type = OurSqlBuilder.class, 
      method = "buildGetByEntityQuery", 
      providerSqlSource = MyProviderSqlSource.class)
  List<T> getByEntity(T entity);
```
If this way can, I will close this PR and resubmit.

Reply:
@h3adache I submitted a new PR: https://github.com/mybatis/mybatis-3/pull/1120